gpii/node_modules/ontologyHandler/src/ontologyHandler.js

```diff
@@ -108,6 +123,156 @@ URLs - this aspect is taken care of in the preferences server component.
         });
         that.options.ontologyTransformSpecs = transformSpecs;
+        // Generate Ontology Routes
+        var edges = [];
+        fluid.each(transformSpecs, function(spec, key) {
+            // The transforms are bidirectional so we add both directed edges
+            var mapping = gpii.ontologyHandler.parseOntologyMapping(key);
+            edges.push(mapping);
+            edges.push({
+                source: mapping.target,
+                target: mapping.source
+            });
+        });
+        that.options.ontologyTransformRoutes = 
+            gpii.ontologyHandler.floydWarshallAlgorithmWithReconstruction(edges);
+    };
+
+    /**
+     * Implementation of the Floyd-Warshall algorithm to determine the shortest
+     * paths between a set of named edges. This is mostly generic, but currently
+     * used in generating a path between multiple linked ontologies.
+     * https://en.wikipedia.org/wiki/Floyd–Warshall_algorithm#Path_reconstruction
+     * Currently assumes that all edges are weight 1.
+     *
+     * @edges (Array) - An array of arrays of length 2, consisting of the named
+     *     edges such as [['A','B'], ['B','C']]
+     * @returns (Object) - An object containing the necessary structures to resolve
+     *     paths using the floydWarshallPath function. Contains 4 members:
+     *     vertices (Object) key/value pair where key is the name of an ontology, and value
+     *         is an integer to be used in the algorithm marking it as a vertice
+     *     numVertices (Integer) Convenience variable for the total number of vertices(ontologies).
+     *     dist (2D Array) dist and next are matrices used in the algorithm, see 
+     *         documentation link above for algorithm details.
+     *     next (2D Array) see dist above
+     */
+    gpii.ontologyHandler.floydWarshallAlgorithmWithReconstruction = function ( edges ) {
+        var data = {};
+        gpii.ontologyHandler.floydWarshallAlgorithmInitVertices(data, edges);
+        gpii.ontologyHandler.floydWarshallAlgorithmInitMatrices(data, edges);
+        gpii.ontologyHandler.floydWarshallAlgorithmCore(data.dist, data.next, data.numVertices);
+        return data;
+    };
+
+    /**
+     * Internal Use only
+     * Function used in gpii.ontologyHandler.floydWarshallAlgorithmWithReconstruction
+     * to initialize the mapping of vertices named by ontology to integers used in
+     * the algorithm.
+     *
+     * @data (Object) Data structure used in main function.
+     * @edges (Array) List of edges where each is an Object contains a directed source and
+     *      target parameter.
+     * @returns Does not a return a value, operates directly on the data structure.
+     */
+    gpii.ontologyHandler.floydWarshallAlgorithmInitVertices = function(data, edges) {
+        data.vertices = {};
+        fluid.each(edges, function(edge) {
+            data.vertices[edge.source] = true;
```

Comment:
What's the second purpose of the structure?  Would it be better if I padded the entries with 1 instead of 'true'?

The purpose of the structure is map a set of integers to the vertices to keep track of them:
{ one: 1, two: 2, three: 3 }

The first step is merely to force the keys to exist so I can fill them.  Otherwise if I did it in one loop it will fill as as:
{ one: 1, two: 3, three: 4}

I guess I could put a counter outside the hash, and only assign numbers if they don't exist already.

```
var count = 1;
fluid.each(edges, function(edge) {
    if (!data.vertices[edge.source]) { data.vertices[edge.source] = count; count += 1; }
    if (!data.vertices[edge.target]) { data.vertices[edge.target] = count; count += 1; }
});